File: nova_lang/lexer.py
"""
NovaLang Regex Lexer with line/column tracking
"""

import re
from .token import Token, TokenType

class LexerError(Exception):
    pass

class Lexer:
    token_spec = [
        ('START', r'start\b'),
        ('END', r'end\b'),
        ('SHOW', r'show\b'),
        ('TAKE', r'take\b'),
        ('WHEN', r'when\b'),
        ('ELSEWHEN', r'elsewhen\b'),
        ('ELSE', r'else\b'),
        ('LOOP', r'loop\b'),
        ('BREAK', r'break\b'),
        ('FUNC', r'func\b'),
        ('BACK', r'back\b'),
        ('NUMKW', r'num\b'),
        ('TEXTKW', r'text\b'),
        ('FLAGKW', r'flag\b'),
        ('TRUE', r'true\b'),
        ('FALSE', r'false\b'),
        ('TO', r'to\b'),

        ('NUMBER', r'\d+'),
        ('STRING', r'"[^"]*"'),
        ('IDENT', r'[A-Za-z_][A-Za-z0-9_]*'),

        ('EQEQ', r'=='),
        ('NOTEQ', r'!='),
        ('GTEQ', r'>='),
        ('LTEQ', r'<='),
        ('GT', r'>'),
        ('LT', r'<'),
        ('ASSIGN', r'='),
        ('PLUS', r'\+'),
        ('MINUS', r'-'),
        ('STAR', r'\*'),
        ('SLASH', r'/'),
        ('COMMA', r','),
        ('LPAREN', r'\('),
        ('RPAREN', r'\)'),
        ('LBRACE', r'\{'),
        ('RBRACE', r'\}'),

        ('SKIP', r'[ \t]+'),
        ('NEWLINE', r'\n'),
        ('MISMATCH', r'.')
    ]

    tok_regex = '|'.join(f'(?P<{name}>{regex})' for name, regex in token_spec)

    def __init__(self, text: str):
        self.text = text
        self.line = 1
        self.col = 1
        self.pos = 0
        self.regex = re.compile(self.tok_regex)

        self.type_map = {
            'START': TokenType.START,
            'END': TokenType.END,
            'SHOW': TokenType.SHOW,
            'TAKE': TokenType.TAKE,
            'WHEN': TokenType.WHEN,
            'ELSEWHEN': TokenType.ELSEWHEN,
            'ELSE': TokenType.ELSE,
            'LOOP': TokenType.LOOP,
            'BREAK': TokenType.BREAK,
            'FUNC': TokenType.FUNC,
            'BACK': TokenType.BACK,
            'NUMKW': TokenType.NUM,
            'TEXTKW': TokenType.TEXT,
            'FLAGKW': TokenType.FLAG,
            'TRUE': TokenType.BOOL,
            'FALSE': TokenType.BOOL,
            'NUMBER': TokenType.NUMBER,
            'STRING': TokenType.STRING,
            'IDENT': TokenType.IDENT,
            'EQEQ': TokenType.EQEQ,
            'NOTEQ': TokenType.NOTEQ,
            'GTEQ': TokenType.GTEQ,
            'LTEQ': TokenType.LTEQ,
            'GT': TokenType.GT,
            'LT': TokenType.LT,
            'ASSIGN': TokenType.ASSIGN,
            'PLUS': TokenType.PLUS,
            'MINUS': TokenType.MINUS,
            'STAR': TokenType.STAR,
            'SLASH': TokenType.SLASH,
            'COMMA': TokenType.COMMA,
            'LPAREN': TokenType.LPAREN,
            'RPAREN': TokenType.RPAREN,
            'LBRACE': TokenType.LBRACE,
            'RBRACE': TokenType.RBRACE,
            'TO': TokenType.TO
        }

    def generate_tokens(self):
        for m in self.regex.finditer(self.text):
            kind = m.lastgroup
            value = m.group()
            if kind == 'NEWLINE':
                self.line += 1
                self.col = 1
                continue
            elif kind == 'SKIP':
                self.col += len(value)
                continue
            elif kind == 'MISMATCH':
                raise LexerError(f"Unexpected character {value!r} at line {self.line}, col {self.col}")
            else:
                ttype = self.type_map[kind]
                token = Token(ttype, value, self.line, self.col)
                yield token
                self.col += len(value)
        yield Token(TokenType.EOF, '', self.line, self.col)

File: nova_lang/lexer.cpp
#include "lexer.hpp"
#include <cctype>
#include <stdexcept>
#include <unordered_map>

Lexer::Lexer(std::string s) : text(std::move(s)), pos(0), line(1), col(1) {}

char Lexer::peek() const { return pos < text.size() ? text[pos] : '\0'; }
char Lexer::peekNext() const { return pos + 1 < text.size() ? text[pos+1] : '\0'; }

char Lexer::advanceChar() {
    char c = peek();
    pos++;
    if (c == '\n') { line++; col = 1; }
    else col++;
    return c;
}

void Lexer::addToken(std::vector<Token>& toks, TokenType ttype, const std::string& val, int l, int c) {
    toks.emplace_back(ttype, val, l, c);
}

bool Lexer::startsWith(const std::string& s) const {
    if (pos + s.size() > text.size()) return false;
    return text.compare(pos, s.size(), s) == 0;
}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> toks;
    static const std::unordered_map<std::string, TokenType> kw = {
        {"start", TokenType::START}, {"end", TokenType::END}, {"show", TokenType::SHOW},
        {"take", TokenType::TAKE}, {"when", TokenType::WHEN}, {"elsewhen", TokenType::ELSEWHEN},
        {"else", TokenType::ELSE}, {"loop", TokenType::LOOP}, {"break", TokenType::BREAK},
        {"func", TokenType::FUNC}, {"back", TokenType::BACK}, {"num", TokenType::NUM},
        {"text", TokenType::TEXT}, {"flag", TokenType::FLAG}, {"to", TokenType::TO},
        {"true", TokenType::BOOL}, {"false", TokenType::BOOL}
    };

    while (pos < text.size()) {
        char c = peek();
        if (c == ' ' || c == '\t' || c == '\r') { advanceChar(); continue; }
        if (c == '\n') { advanceChar(); continue; }
        if (c == '#') { // comment
            while (peek() != '\n' && peek() != '\0') advanceChar();
            continue;
        }

        int start_line = line, start_col = col;

        // Numbers
        if (std::isdigit(c)) {
            std::string num;
            while (std::isdigit(peek())) num.push_back(advanceChar());
            addToken(toks, TokenType::NUMBER, num, start_line, start_col);
            continue;
        }

        // Identifiers / keywords / bools
        if (std::isalpha(c) || c == '_') {
            std::string id;
            while (std::isalnum(peek()) || peek() == '_') id.push_back(advanceChar());
            std::string lower = id;
            for (auto& ch : lower) ch = (char)std::tolower(ch);
            auto it = kw.find(lower);
            if (it != kw.end()) {
                TokenType tt = it->second;
                if (tt == TokenType::BOOL) {
                    addToken(toks, TokenType::BOOL, lower, start_line, start_col);
                } else {
                    addToken(toks, tt, lower, start_line, start_col);
                }
            } else {
                addToken(toks, TokenType::IDENT, id, start_line, start_col);
            }
            continue;
        }

        // Strings
        if (c == '"') {
            advanceChar(); // consume "
            std::string s;
            while (peek() != '"' && peek() != '\0') {
                char ch = advanceChar();
                if (ch == '\\' && peek() != '\0') {
                    char esc = advanceChar();
                    if (esc == 'n') s.push_back('\n');
                    else if (esc == '"') s.push_back('"');
                    else s.push_back(esc);
                } else s.push_back(ch);
            }
            if (peek() != '"') throw std::runtime_error("Unterminated string");
            advanceChar(); // consume closing "
            addToken(toks, TokenType::STRING, s, start_line, start_col);
            continue;
        }

        // Two-char ops
        if (startsWith("==")) { addToken(toks, TokenType::EQEQ, "==", start_line, start_col); advanceChar(); advanceChar(); continue; }
        if (startsWith("!=")) { addToken(toks, TokenType::NOTEQ, "!=", start_line, start_col); advanceChar(); advanceChar(); continue; }
        if (startsWith(">=")) { addToken(toks, TokenType::GTEQ, ">=", start_line, start_col); advanceChar(); advanceChar(); continue; }
        if (startsWith("<=")) { addToken(toks, TokenType::LTEQ, "<=", start_line, start_col); advanceChar(); advanceChar(); continue; }

        // Single char ops/punct
        switch (c) {
            case '+': addToken(toks, TokenType::PLUS, "+", start_line, start_col); advanceChar(); break;
            case '-': addToken(toks, TokenType::MINUS, "-", start_line, start_col); advanceChar(); break;
            case '*': addToken(toks, TokenType::STAR, "*", start_line, start_col); advanceChar(); break;
            case '/': addToken(toks, TokenType::SLASH, "/", start_line, start_col); advanceChar(); break;
            case '=': addToken(toks, TokenType::ASSIGN, "=", start_line, start_col); advanceChar(); break;
            case '>': addToken(toks, TokenType::GT, ">", start_line, start_col); advanceChar(); break;
            case '<': addToken(toks, TokenType::LT, "<", start_line, start_col); advanceChar(); break;
            case ',': addToken(toks, TokenType::COMMA, ",", start_line, start_col); advanceChar(); break;
            case '(' : addToken(toks, TokenType::LPAREN, "(", start_line, start_col); advanceChar(); break;
            case ')' : addToken(toks, TokenType::RPAREN, ")", start_line, start_col); advanceChar(); break;
            case '{' : addToken(toks, TokenType::LBRACE, "{", start_line, start_col); advanceChar(); break;
            case '}' : addToken(toks, TokenType::RBRACE, "}", start_line, start_col); advanceChar(); break;
            default:
                {
                    std::string msg = "Unexpected character: ";
                    msg.push_back(c);
                    throw std::runtime_error(msg);
                }
        }
    }

    addToken(toks, TokenType::EOF_T, "", line, col);
    return toks;
}
